drivers/docker/driver.go

```diff
@@ -0,0 +1,1250 @@
+		return fmt.Errorf("command contains extra white space: %q", command)
+	}
+
+	return nil
+}
+
+func (d *Driver) WaitTask(ctx context.Context, taskID string) (<-chan *drivers.ExitResult, error) {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return nil, drivers.ErrTaskNotFound
+	}
+	ch := make(chan *drivers.ExitResult)
+	go d.handleWait(ctx, ch, h)
+	return ch, nil
+}
+
+func (d *Driver) handleWait(ctx context.Context, ch chan *drivers.ExitResult, h *taskHandle) {
+	defer close(ch)
+	select {
+	case <-h.waitCh:
+		ch <- h.exitResult
+	case <-ctx.Done():
+		ch <- &drivers.ExitResult{
+			Err: ctx.Err(),
+		}
+	}
+}
+
+func (d *Driver) StopTask(taskID string, timeout time.Duration, signal string) error {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return drivers.ErrTaskNotFound
+	}
+
+	if signal == "" {
+		signal = "SIGINT"
+	}
+
+	sig, err := signals.Parse(signal)
+	if err != nil {
+		return fmt.Errorf("failed to parse signal: %v", err)
+	}
+
+	return h.Kill(timeout, sig)
+}
+
+func (d *Driver) DestroyTask(taskID string, force bool) error {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return drivers.ErrTaskNotFound
+	}
+
+	c, err := h.client.InspectContainer(h.containerID)
+	if err != nil {
+		return fmt.Errorf("failed to inspect container state: %v", err)
+	}
+	if c.State.Running && !force {
+		return fmt.Errorf("must call StopTask for the given task before Destroy or set force to true")
+	}
+
+	defer h.dloggerPluginClient.Kill()
+	if err := h.client.StopContainer(h.containerID, 0); err != nil {
+		h.logger.Warn("failed to stop container during destroy", "error", err)
+	}
+
+	if err := h.dlogger.Stop(); err != nil {
+		h.logger.Error("failed to stop docker logger process during destroy",
+			"error", err, "logger_pid", h.dloggerPluginClient.ReattachConfig().Pid)
+	}
+
+	if err := d.cleanupImage(h); err != nil {
+		h.logger.Error("failed to cleanup image after destroying container",
+			"error", err)
+	}
+
+	return nil
+}
+
+// cleanupImage removes a Docker image. No error is returned if the image
+// doesn't exist or is still in use. Requires the global client to already be
+// initialized.
+func (d *Driver) cleanupImage(handle *taskHandle) error {
+	if !d.config.GC.Image {
+		return nil
+	}
+
+	d.coordinator.RemoveImage(handle.containerImage, handle.task.ID)
+
+	return nil
+}
+
+func (d *Driver) InspectTask(taskID string) (*drivers.TaskStatus, error) {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return nil, drivers.ErrTaskNotFound
+	}
+
+	container, err := client.InspectContainer(h.containerID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to inspect container %q: %v", h.containerID, err)
+	}
+	status := &drivers.TaskStatus{
+		ID:          h.task.ID,
+		Name:        h.task.Name,
+		StartedAt:   container.State.StartedAt,
+		CompletedAt: container.State.FinishedAt,
+		DriverAttributes: map[string]string{
+			"container_id": container.ID,
+		},
+		NetworkOverride: h.net,
+		ExitResult:      h.exitResult,
```

Comment:
InspectTask may be called concurrently with `taskHandle.run()` exiting, so I think both sides need to lock around the shared fields.